src/memspaces/memspace_host.c

```diff
@@ -0,0 +1,101 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include <assert.h>
+#include <hwloc.h>
+#include <stdlib.h>
+
+#include "base_alloc_global.h"
+#include "memory_target_numa.h"
+#include "memspace_internal.h"
+#include "memspace_numa.h"
+
+static umf_result_t umfMemspaceHostCreate(umf_memspace_handle_t *hMemspace) {
+    if (!hMemspace) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_result_t umf_ret = UMF_RESULT_SUCCESS;
+
+    hwloc_topology_t topology;
+    if (hwloc_topology_init(&topology)) {
+        // TODO: What would be an approrpiate err?
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    if (hwloc_topology_load(topology)) {
+        umf_ret = UMF_RESULT_ERROR_UNKNOWN;
+        goto err_topology_destroy;
+    }
+
+    // Shouldn't return -1, as 'HWLOC_OBJ_NUMANODE' deosn't appear to be an
+    // object that can be present on multiple levels.
+    // Source: https://www.open-mpi.org/projects/hwloc/doc/hwloc-v2.10.0-letter.pdf
+    size_t nNodes =
+        (size_t)hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_NUMANODE);
+
+    umf_ba_pool_t *base_allocator = umf_ba_get_pool(nNodes * sizeof(size_t));
+    if (!base_allocator) {
+        umf_ret = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_topology_destroy;
+    }
+
+    size_t *nodeIds = (size_t *)umf_ba_alloc(base_allocator);
+    if (!nodeIds) {
+        umf_ret = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_topology_destroy;
+    }
+
+    // Collect all available NUMA node ids on the platform
+    size_t nodeIdx = 0;
+    hwloc_obj_t numaNodeObj = NULL;
+    while ((numaNodeObj = hwloc_get_next_obj_by_type(
+                topology, HWLOC_OBJ_NUMANODE, numaNodeObj)) != NULL) {
+        // Shouldn't be possible to iterate over more 'HWLOC_OBJ_NUMANODE' objects
+        // than the number returned by hwloc_get_nbobjs_by_type.
+        assert(nodeIdx < nNodes);
+        nodeIds[nodeIdx++] = numaNodeObj->os_index;
+    }
+
+    umf_ret = umfMemspaceCreateFromNumaArray(nodeIds, nNodes, hMemspace);
+
+    umf_ba_free(base_allocator, nodeIds);
+
+err_topology_destroy:
+    hwloc_topology_destroy(topology);
+    return umf_ret;
+}
+
+static bool UMF_MEMSPACE_HOST_INITIALIZED = false;
+static umf_memspace_handle_t UMF_MEMSPACE_HOST = NULL;
+
+umf_result_t umfMemspaceHostGet(umf_memspace_handle_t *hMemspace) {
+    if (!hMemspace) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!UMF_MEMSPACE_HOST_INITIALIZED) {
```

Comment:
can't you just check if UMF_MEMSPACE_HOST == NULL ?

Reply:
Done.